Source/WebCore/platform/audio/AudioDestinationResampler.cpp

```diff
@@ -0,0 +1,148 @@
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "AudioDestinationResampler.h"
+
+#if ENABLE(WEB_AUDIO)
+
+#include "AudioBus.h"
+#include "AudioSession.h"
+#include "AudioUtilities.h"
+#include "Logging.h"
+#include "MultiChannelResampler.h"
+
+namespace WebCore {
+
+constexpr size_t fifoSize = 96 * AudioUtilities::renderQuantumSize;
+
+AudioDestinationResampler::AudioDestinationResampler(AudioIOCallback& callback, unsigned numberOfOutputChannels, float inputSampleRate, float outputSampleRate)
+    : AudioDestination(callback, inputSampleRate)
+    , m_outputBus(AudioBus::create(numberOfOutputChannels, AudioUtilities::renderQuantumSize, false).releaseNonNull())
+    , m_renderBus(AudioBus::create(numberOfOutputChannels, AudioUtilities::renderQuantumSize).releaseNonNull())
+    , m_fifo(numberOfOutputChannels, fifoSize)
+{
+    if (inputSampleRate != outputSampleRate) {
+        double scaleFactor = static_cast<double>(inputSampleRate) / outputSampleRate;
+        m_resampler = makeUnique<MultiChannelResampler>(scaleFactor, numberOfOutputChannels, AudioUtilities::renderQuantumSize, [this](AudioBus* bus, size_t framesToProcess) {
+            ASSERT_UNUSED(framesToProcess, framesToProcess == AudioUtilities::renderQuantumSize);
+            callRenderCallback(nullptr, bus, AudioUtilities::renderQuantumSize, m_outputTimestamp);
+        });
+    }
+}
+
+AudioDestinationResampler::~AudioDestinationResampler() = default;
+
+unsigned AudioDestinationResampler::framesPerBuffer() const
+{
+    return m_renderBus->length();
+}
+
+void AudioDestinationResampler::start(Function<void(Function<void()>&&)>&& dispatchToRenderThread, CompletionHandler<void(bool)>&& completionHandler)
+{
+    ASSERT(isMainThread());
+    LOG(Media, "AudioDestinationResampler::start");
+    {
+        Locker locker { m_dispatchToRenderThreadLock };
+        m_dispatchToRenderThread = WTFMove(dispatchToRenderThread);
+    }
+    startRendering(WTFMove(completionHandler));
+}
+
+void AudioDestinationResampler::stop(CompletionHandler<void(bool)>&& completionHandler)
+{
+    ASSERT(isMainThread());
+    LOG(Media, "AudioDestinationResampler::stop");
+    stopRendering(WTFMove(completionHandler));
+    {
+        Locker locker { m_dispatchToRenderThreadLock };
+        m_dispatchToRenderThread = nullptr;
+    }
+}
+
+void AudioDestinationResampler::setIsPlaying(bool isPlaying)
+{
+    ASSERT(isMainThread());
+
+    if (m_isPlaying == isPlaying)
+        return;
+
+    m_isPlaying = isPlaying;
+
+    {
+        Locker locker { m_callbackLock };
+        if (m_callback)
+            m_callback->isPlayingDidChange();
+    }
+}
+
+size_t AudioDestinationResampler::pullRendered(size_t numberOfFrames)
+{
+    ASSERT(!isMainThread());
+    ASSERT(numberOfFrames <= fifoSize);
+    Locker locker { m_fifoLock };
+    return m_fifo.pull(m_outputBus.ptr(), numberOfFrames);
```

Comment:
It is good to assert when asked to pull too many frames in a debug build, but what about in a release build? Should we RELEASE_ASSERT, or clamp `numberOfFrames` to  `fifoSize`?